exporter/otlphttpexporter/otlp.go

```diff
@@ -199,29 +203,102 @@ func isRetryableStatusCode(code int) bool {
 	}
 }
 
+func readResponseBody(body io.ReadCloser) ([]byte, error) {
+	// Read the maximum number of bytes allowed in a request. This avoids
+	// issues with missing or invalid Content-Length headers.
+	protoBytes := make([]byte, maxHTTPResponseReadBytes)
+	n, err := io.ReadFull(body, protoBytes)
+
+	// No bytes read and an EOF error indicates there is no body to read.
+	if n == 0 && (err == nil || errors.Is(err, io.EOF)) {
+		return nil, nil
+	}
+
+	// io.ReadFull will return io.ErrorUnexpectedEOF in most cases since there
+	// will usually be a mismatch between the length of the byte slice and the
+	// size of the body, so we ignore that error.
+	if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return nil, err
+	}
+
+	// The pdata unmarshaling methods check for the length of the slice
+	// when unmarshaling it, so we have to trim down the length to the
+	// actual size of the data.
+	return protoBytes[:n], nil
+}
+
 // Read the response and decode the status.Status from the body.
 // Returns nil if the response is empty or cannot be decoded.
-func readResponse(resp *http.Response) *status.Status {
+func readResponseStatus(resp *http.Response) *status.Status {
 	var respStatus *status.Status
 	if resp.StatusCode >= 400 && resp.StatusCode <= 599 {
 		// Request failed. Read the body. OTLP spec says:
 		// "Response body for all HTTP 4xx and HTTP 5xx responses MUST be a
 		// Protobuf-encoded Status message that describes the problem."
-		maxRead := resp.ContentLength
-		if maxRead == -1 || maxRead > maxHTTPResponseReadBytes {
-			maxRead = maxHTTPResponseReadBytes
```

Comment:
Why is this logic dropped in `readResponseBody`? I don't think we want to allocate maxHTTPResponseReadBytes's long slices every time

Reply:
This was removed in response to earlier feedback: https://github.com/open-telemetry/opentelemetry-collector/pull/6970#discussion_r1143533440. The logic for determining the correct slice size is a little complicated, and 64 KiB doesn't feel like a substantial amount of memory. If there's consensus on allocating smaller slices, I'm fine with that. We could also allocate a single maxHTTPResponseReadBytes-sized slice (or pool of slices) within this component and reuse it if that would be preferable.

Reply:
What's the typical OTLP response size?

I think the comment was about introducing `needsResize` not calculating the slice size. If the typical OTLP response size is much smaller than 64k, I don't think we should always allocate 64k.

Reply:
I would estimate that for most cases it's well below 1 KiB, the largest responses will likely be partial success responses with details in the message. I'm also fine with saving memory where we can, I've added back the logic to use the Content-Length header to size the slice.

I re-read the comment and you're right, I think I misread it. Providing a smaller slice through re-slicing is an extremely cheap operation anyway now that I'm thinking of it, so I've removed the `needsResize` variable.